torchvision/prototype/transforms/functional/_color.py

```diff
@@ -289,7 +295,21 @@ def adjust_gamma(inpt: features.InputTypeJIT, gamma: float, gain: float = 1) ->
         return adjust_gamma_image_pil(inpt, gamma=gamma, gain=gain)
 
 
-posterize_image_tensor = _FT.posterize
+def posterize_image_tensor(image: torch.Tensor, bits: int) -> torch.Tensor:
+    if image.is_floating_point():
+        levels = 1 << bits
+        return image.mul(levels).floor_().clamp_(0, levels - 1).div_(levels)
```

Comment:
Can you explain the clamp at levels-1? [This](https://www.shadertoy.com/view/wlBBRy) is the kind of implementation reference I had in mind. Which reference did you use?

Reply:
I touched on this in 3. in my top comment. Since the input range for float images is inclusive on the edges, i.e. `[0.0, 1.0]`, `image.mul(levels).floor_()` gives us `levels + 1` values, i.e. `{0.0, 1.0, 2.0, ..., levels - 1.0, levels}`. 

However, we want the kernel to quantize to `levels` levels. Thus, we need to remove one level. For integer dtypes, the higher values are removed, i.e. the remaining values are `{i * 2 ** (bit_depth - bits) for i in range(2 ** bits)}`. For example

```pycon
>>> bits = 3
>>> bit_depth = 8
>>> {i * 2 ** (bit_depth - bits) for i in range(2 ** bits)}
{0, 32, 64, 96, 128, 160, 192, 224}
```

As you can see the `255` / `256` corresponding to `1.0` in floating point images is missing. Thus, we also clamp that away for floating images.

Reply:
I read your PR description but I still had questions on this. Why not do `image.mul(levels-1)` to begin with? Multiplying by `levels` means that upper bound of `1` will go outside of the permitted range of the type. What am I missing here?

Reply:
To make this more concrete, let's look at an example:

```pycon
>>> image = torch.arange(0, 256, dtype=torch.uint8)
>>> bits = 3
>>> output_baseline = image & (2 ** 8 - 2 ** (8 - bits))
>>> torch.unique(output_baseline)
tensor([  0,  32,  64,  96, 128, 160, 192, 224], dtype=torch.uint8)
```

```pycon
>>> image = torch.linspace(0, 1, 100)
>>> output1 = image.mul(2 ** bits).floor_().clamp_(0, 2**bits - 1).div_(2 ** bits)
>>> torch.unique(output.mul(255).byte())
tensor([  0,  31,  63,  95, 127, 159, 191, 223], dtype=torch.uint8)
```

The proposal in this is not perfect, but the `.byte()` call above eliminates some nuances. The diff is actually `<1`.

In contrast if we do what you propose we get

```pycon
>>> output2 = image.mul(2 ** bits - 1).floor_().div(2 ** bits - 1)
>>> torch.unique(output2.mul(255).byte())
tensor([  0,  36,  72, 109, 145, 182, 218, 255], dtype=torch.uint8)
```

This is of course also a valid way to posterize an image, but the behavior is divergent from what we and PIL do for integers.

Reply:
Thanks for the explanation @pmeier. This makes sense. @vfdev-5 thoughts?

Reply:
@pmeier I think you can skip clamp with the following:
```
(x * (2 ** bits - 1)).floor() / (2 ** bits)
```

Reply:
Unfortunately, this does not work. While the individual level values look good

```pycon
>>> output3 = (image * (2 ** bits - 1)).floor() / (2 ** bits)
>>> torch.unique(output3.mul(255).byte())
tensor([  0,  31,  63,  95, 127, 159, 191, 223], dtype=torch.uint8)
```

the posterized values do not match what we do for the integers:

```py
>>> _ = torch.manual_seed(0)
>>> bits = 3
>>> image_uint8 = torch.randint(0, 256, (3, 3), dtype=torch.uint8)
>>> posterized_image_uint8 = image_uint8 & (2 ** 8 - 2 **(8 - bits))
>>> posterized_image_uint8
tensor([[160,  32,  96],
        [192,  64, 224],
        [192,  96,   0]], dtype=torch.uint8)
>>> image_float32 = F.convert_dtype_image_tensor(image_uint8, torch.float32)
>>> posterized_image_float32 = (image_float32 * (2 ** bits - 1)).floor() / (2 ** bits)
>>> posterized_image_float32 = posterized_image_float32.mul(255).byte()
>>> posterized_image_float32
tensor([[127,  31,  95],
        [159,  31, 191],
        [159,  63,   0]], dtype=torch.uint8)
>>> posterized_image_uint8.int() - posterized_image_float32.int()
tensor([[33,  1,  1],
        [33, 33, 33],
        [33, 33,  0]], dtype=torch.int32)
```

Reply:
Re EDIT 2: formula still produces different values:

```pycon
>>> posterized_image_float32 = (image_float32 * (2 ** bits - 0.5)).floor() / (2 ** bits)
>>> posterized_image_float32.mul(255).byte()
tensor([[159,  31,  95],
        [159,  31, 223],
        [159,  95,   0]], dtype=torch.uint8)
>>> posterized_image_uint8.int() - posterized_image_float32.int()
tensor([[ 1,  1,  1],
        [33, 33,  1],
        [33,  1,  0]], dtype=torch.int32)
```

Reply:
Well, using clamp you still have a difference of +/- 1 but OK, let's have clamp. Probably, it is not a big deal in terms of runtime perfs.